### DISCUSSION TITLE
Implement restrictions on creating multiple challenges to prevent spam and exploitation of rewards

### DISCUSSION THREAD

**Discussion by behnamazimi** [sureyoudo-io/SureYouDo.io-smart-contracts#8] [General]
Currently, a single wallet can create multiple challenges within a very short period, leading to potential spamming issues. There is a concern that users might exploit the reward system, as finalizing a challenge grants rewards in SYD. Some may create challenges with minimal pledges primarily to collect these rewards.

Although a challenge requires at least three transactions to get SYD rewards (creation, verification, and finalization for challenges with an end goal only), the potential for misuse still exists. To safeguard the system, it would be safe to consider implementing preventive measures in the smart contract.

**Discussion Points:**

- What are effective strategies to integrate into the smart contract to prevent this type of spamming?
- How can we ensure that these measures are robust enough to accommodate future changes and challenges?

Your insights on this would be greatly appreciated.

**Comment by behnamazimi** [comment]
Would appreciate it if you could also share your thoughts @rohallah12

**Comment by 0xScratch** [comment]
That's a good set of questions as well as a discussion point that might brought up certain ideas to prevent from spamming..
Right now, as I was reading the main description by @behnamazimi two ideas struck my mind that doesn't needed much thinking and could easily work out...

1. Using **Number of Days** as a limiter: We can include a specific limit that doesn't let anyone create a challenge less than that particular days.
2. Using **AI** as a goal checker: The AI will check whether the goal seems valid or not, and it also check that goal shouldn't be the one which gets completed in a few span of time.
3. We can even use a combination of `1` and `2`

Ideas still brewing, will love to post them here as got something interesting in mind...Wondering what @behnamazimi thinks about this problem?

**Comment by behnamazimi** [comment]
I prefer an on-chain solution, which makes me think of implementing it as follows:

We can check if some time has passed since the latest challenge was created by a specific address. We already have the data for this ((ref)[https://github.com/sureyoudo-io/SureYouDo.io-smart-contracts/blob/main/contracts/SydChallengeManager.sol#L41]) ; we can use the 'createdAt' field of each challenge. 

We can add another config state like `challengeCreationCooldownPeriod` that stores value in minutes and we use that to check the passed time.

Or, we can add another state to keep the timestamp of the last created challenge and compare that with passed time. 

But I'm not sure about its gas efficiency. If you have any other gas-efficient ways to implement this, please suggest.

**Reply by 0xScratch** [reply]
There's one thing to ask here, Like suppose I am creating a challenge of `Learning Rust` let's say...this makes two ways to go, one is you are asking the person itself what days he thing will need that challenge to complete and will set the timer according to that, which won't really prevent that problem as I can easily set 2 days and get my rewards. Another way is, you are hard-coding the dapp in order to set it in this way - No, challenge will be created with time less than 3 days (let's say).
So, what's way you are thinking it to be?

**Reply by behnamazimi** [reply]
The minimum duration for each challenge is from the current time until the end of the day (UTC). However, the duration itself is not the primary concern here. What we're focused on is, for instance, if you create multiple (like 100) challenges for 'Learning Rust' and all of them have the end date set for today. In such cases, you can easily claim the rewards at the end of the day.

The objective here is to prevent such scenarios by introducing a cooldown period between creating challenges. For example, if we implement a cooldown period of 6 hours, each account will only be able to create a maximum of 4 challenges.

Does it make sense?

**Reply by 0xScratch** [reply]
Hmm, I got it somehow but now that makes me wonder a common scenario...Like we usually create our to-do lists sometimes and introduce all our tasks at one go for either our day or whatever timeline, Probably the person wouldn't want to wait for another challenge until 6 hours cooldown (if our cooldown period is of 6 hours). It be like we are really limiting up the challenges he can create at a time...right?

Started feeling like we have to tweak the reward system in a way that our users don't really get focused much on rewards but more on their personal improvements which leads them to honest actions.

(Now this discussion getting on a interesting path!)

**Reply by behnamazimi** [reply]
This a good point. Let me think about it.

**Reply by behnamazimi** [reply]
@0xScratch what if we add a reward limit per wallet per day instead of the cooldown period?

**Reply by 0xScratch** [reply]
@behnamazimi I don't think that's a good way to do, cuz then it will limit the user's action. Suppose I created some challenges and did them with honesty and got some rewards for it, works fine...but if I still have time left in my day but the reward limit is over, although I still can do challenges but due to the greed of those rewards I will postpone them. Actually, we have to think here on behalf of the user's perspective as well as the system's one.

I am thinking of an idea that can even be wrong but still didn't think how...anyways let me share it with you.
What if we provide rewards based on the challenge's duration, that is challenges with less duration will be provided with less rewards than the ones with longer duration. As we will prefer long-term challenges like `Going Gym continuous for a month`, `Learning Rust` rather than short-term like `Attending my music class tomorrow`, `Baking a cake`.
Also, there's one more idea which seems quite complex but sounds worthy...creating a system or an algorithm that detects suspicious pattern. We might make it work like -> Rewards won't be given at instant but be sent after sometime (let's say 1 day). With that the algo will check the challenges completed in a given time based on a possible ratio, like completing 4-5 challenges in 2 days seems normal and acceptable but completing 40-50 challenges in 2 days doesn't as it clearly seems to be a suspicious activity. Yea, this idea is complex but I feel this will work better.
Also, was thinking of why not dividing up the challenges in parts i.e providing them rewards on progress rather than in one go

Let's see what your say on this!

**Reply by behnamazimi** [reply]
> I don't think that's a good way to do, cuz then it will limit the user's action

The point here is that you're considering short-term challenges, but I'm focusing more on challenges with longer durations. You're discussing users who use the app as their to-do management platform, but that's not actually the intended purpose of the platform. So, for me, it's an edge case that we can address in the long term. 

> What if we provide rewards based on the challenge's duration

We can't measure the true value of challenges solely based on their durations. For instance, how can we compare "building a nice project in 4 weeks" with "going to the gym for 4 weeks"? I really don't think we should approach it this way.

> creating a system or an algorithm that detects suspicious pattern

I like this idea, although perhaps we could implement it in the next phases as it's a bit challenging and needs centralized data storage for the platform we can query on. 

Currently, by having a daily reward limit per wallet, we can effectively prevent spam. Imagine if we could configure that limit based on user feedback and the latest momentum. And, in the long term, we could combine your idea of implementing an algorithm with gradually increasing the rewards-per-day limits.

Does that make any sense?

**Reply by 0xScratch** [reply]
>The point here is that you're considering short-term challenges, but I'm focusing more on challenges with longer durations. You're discussing users who use the app as their to-do management platform, but that's not actually the intended purpose of the platform. So, for me, it's an edge case that we can address in the long term.

Well, you are right here. and now that makes more easier to think about these cases as shorter duration challenges won't be accounted for.

>We can't measure the true value of challenges solely based on their durations. For instance, how can we compare "building a nice project in 4 weeks" with "going to the gym for 4 weeks"? I really don't think we should approach it this way.

As shorter-duration challenges aren't taken into account it doesn't make any sense to provide rewards based on duration.

>Currently, by having a daily reward limit per wallet, we can effectively prevent spam. Imagine if we could configure that limit based on user feedback and the latest momentum. And, in the long term, we could combine your idea of implementing an algorithm with gradually increasing the rewards-per-day limits.

So, a daily rewards limit per wallet makes sense in terms of longer duration and yes that limit can be configured the way you said.

**Reply by behnamazimi** [reply]
Cool! I've created an issue #14  for this. Feel free to assign it to yourself if you want to work on it. Otherwise, I'll assign it to myself tomorrow and begin working on it. @0xScratch

**Reply by 0xScratch** [reply]
Actually, I am busy with my college stuff (even if I don't like it). Even discussions at a certain length are quite difficult to manage when you have some exams going on. So, it is better if you assign it to yourself as that seems to be a good option here (as this thing requires quite an attention and time).
Will do some coding stuff after I get some time from this college shit. But do keep pinging up for any new cool features, updates or short talks about the current implementations if you can!

**Reply by behnamazimi** [reply]
Oh! Good luck!

**Comment by behnamazimi** [comment]
> Or, we can add another state to keep the timestamp of the last created challenge and compare that with passed time.

I've just tested this and it consumes more gas fees than the other way - using `_userChallenges`, FYI.

**Comment by behnamazimi** [comment]
I've addressed this in PR https://github.com/sureyoudo-io/SureYouDo.io-smart-contracts/pull/9

**Comment by behnamazimi** [comment]
Final implementation: https://github.com/sureyoudo-io/SureYouDo.io-smart-contracts/issues/14
